Add bloc_library example

felangel opened the pull request.

mmcc007 commented:
Thanks for the PR!

Can you add something like the following to the .travis.yml so that travis will run the integration tests for this package:
```yaml
    - <<: *integration-test-staging
      os: linux
      env: built_redux_android
      script: travis_retry ./scripts/ci.sh ./example/bloc_library
    - <<: *integration-test-staging
      os: osx
      osx_image: xcode8.0
      env: built_redux_ios
      before_install: *before_install_osx
      script: travis_retry ./scripts/ci.sh ./example/bloc_library
```
👍

felangel commented:
@mmcc007 sorry about that! I just made the change 😄

brianegan commented on example/bloc_library/test/todos_bloc_test.dart:
```diff
@@ -0,0 +1,139 @@
+import 'package:bloc_library/models/models.dart';
+import 'package:todos_repository_flutter/todos_repository_flutter.dart';
+
+class MockTodosRepository extends Mock implements TodosRepositoryFlutter {}
```
Beautiful 😻

brianegan commented on example/bloc_library/lib/blocs/todos/todos_bloc.dart:
```diff
@@ -0,0 +1,103 @@
+    }
+  }
+
+  Stream<TodosState> _mapDeleteTodoToState(
```
One error with this app: Please try mutating the todos then doing a Hot Restart. You will see the same todos that were initially loaded from the Mock API are returned once again.


In all of these operations that mutate the list of todos (add, remove, edit, clear completed, etc), you'll need to make a call out to the repository to save the changes.

Then, when the app is restarted, the repository should be up to date and return the correct current list of todos!

This is the one thing I haven't figured out how to integration test yet. Wonder if there's a newer method on the test driver for this!

felangel replied:
Oops! I’ll fix that asap 👍

mmcc007 replied:
FYI: Until such feature is available on flutter_driver, it is possible to auto-restart an app via adb and re-run an integration test (or run another integration test) without reinstalling the app. Try the following for android:
```
#!/usr/bin/env bash
# Originally written by Maurice McCabe <mmcc007@gmail.com>, but placed in the public domain.

# Run Flutter integration tests on android device (or emulator)

show_help() {
    printf "\n\nusage: $0 package test

Utility for running integration tests for pre-installed flutter app on android device.
(app must be built in debug mode with 'enableFlutterDriverExtension()')

where:
    package
        name of package to run, eg, com.example.flutterapp
    test
        path of test to run, eg, test_driver/main_test.dart
"
    exit 1
}

run_test() {
    # note: assumes debug apk installed on device
    package_name=$1
    test_path=$2

    echo Starting Flutter app $package_name in debug mode...

    # stop app on device
    # (if running locally or started incorrectly by CI/CD)
    adb shell am force-stop $package_name

    # clear log (to avoid picking up any earlier observatory announcements on local re-runs)
    # (could comment out if running in CI/CD)
    adb logcat -c

    # start app on device
    adb shell am start -a android.intent.action.RUN -f 0x20000000 --ez enable-background-compilation true --ez enable-dart-profiling true --ez enable-checked-mode true $package_name/$package_name.MainActivity

    # wait for observatory startup on device and get port number
    obs_str=`adb logcat -e "Observatory listening on" -m 1`
    obs_port=`echo $obs_str | grep -Eo '([0-9]+)/$'`
    obs_port=${obs_port%?}
    echo Observatory on port $obs_port

    # forward a local port to observatory port on device
    forwarded_port=`adb forward tcp:0 tcp:$obs_port`
#    forwarded_port=4723 # re-use appium server port for now
    adb forward tcp:$forwarded_port tcp:$obs_port
    echo Local port $forwarded_port forwarded to observatory port $obs_port

    # run test
    echo Running integration test $test_path on app $package_name ...
    flutter packages get # required when running in CI/CD
    export VM_SERVICE_URL=http://127.0.0.1:$forwarded_port
    dart $test_path
#    flutter driver --use-existing-app=http://127.0.0.1:$forwarded_port --no-keep-app-running lib/main.dart

}

# if no arguments passed
if [[ -z $1  || -z $2 ]]; then show_help; fi

run_test $1 $2
```
Should be possible to do something similar for iOS, just haven't got it working yet...

mmcc007 replied:
I posted an issue about flutter driver regarding this:
https://github.com/flutter/flutter/issues/28662

brianegan commented on example/bloc_library/lib/blocs/simple_bloc_delegate.dart:
```diff
@@ -0,0 +1,17 @@
+
+import 'package:bloc/bloc.dart';
+
+class SimpleBlocDelegate extends BlocDelegate {
```
Maybe a quick comment about this class or links to docs? Out of context, it looks a bit weird that you're creating a dummy override then mutating some (perhaps singleton) class with this dummy delegate?

felangel replied:
✅

brianegan commented on example/bloc_library/lib/main.dart:
```diff
@@ -0,0 +1,63 @@
+import 'package:bloc_library/screens/screens.dart';
+
+void main() {
+  BlocSupervisor().delegate = SimpleBlocDelegate();
```
Wat dis do? (Connected to other comment)

felangel replied:
It allows developers to create their own delegate which exposes things like global error handling as well as global transition handling. I’ll definitely update the docs to include more info 💯

felangel commented on example/bloc_library/README.md:
```diff
@@ -0,0 +1,34 @@
+
+This sample makes use of the [bloc](https://pub.dartlang.org/packages/bloc) and [flutter_bloc](https://pub.dartlang.org/packages/flutter_bloc) libraries to manage state.
+
+Check out the [bloc library documentation](https://felangel.github.io/bloc) for more details and tutorials.
```
Great point!

felangel commented:
@brianegan I think I have addressed all of the feedback with the exception of widget tests. Is it alright if you review the PR as is and I can create a separate PR to add them as part of (#119)?

Thanks for all the awesome feedback! 💯

brianegan commented:
Changes look great, thanks!
